http/responseemitter.go

```diff
@@ -112,19 +111,23 @@ func (re *responseEmitter) Emit(value interface{}) error {
 		isSingle = true
 	}
 
+	if f, ok := re.w.(http.Flusher); ok {
+		defer f.Flush()
+	}
+
 	switch v := value.(type) {
+	case error:
+		return re.closeWithError(v)
 	case io.Reader:
 		err = flushCopy(re.w, v)
 	default:
 		err = re.enc.Encode(value)
 	}
 
-	if f, ok := re.w.(http.Flusher); ok {
-		f.Flush()
-	}
-
 	if isSingle {
-		err = re.closeWithError(err)
+		// always close singles with nil error.
+		// encoding errors go to caller.
+		err = re.closeWithError(nil)
```

Comment:
If we return the error to the "caller", expecting them to do something, we *can't* close. This should be something like `isSingle && err != nil`. Am I missing something?

Reply:
It should have been `isSingle && err == nil`. If an encoding or copy error occurs, we didn't send the value so it doesn't make sense to close the emitter but we should return it. If encoding went fine, gracefully close the emitter.